src/psyclone/tests/psyir/transformations/fold_conditional_return_expressions_trans_test.py

```diff
@@ -0,0 +1,162 @@
+
+
+def test_description():
+    ''' Check that the transformation returns the expected strings '''
+    trans = FoldConditionalReturnExpressionsTrans()
+    assert trans.name == "FoldConditionalReturnExpressionsTrans"
+    assert str(trans) == ("Re-structure kernel statements to eliminate "
+                          "conditional Return expressions.")
+
+
+def test_validation():
+    ''' Check that the transformation can only be applied to routine nodes '''
+    trans = FoldConditionalReturnExpressionsTrans()
+    with pytest.raises(TransformationError) as info:
+        trans.apply(None)
+    assert("Error in FoldConditionalReturnExpressionsTrans transformation. "
+           "This transformation can only be applied to 'Routine' nodes, but "
+           "found 'NoneType'." in str(info.value))
+
+
+SUB_IN1 = (
+    "subroutine sub1(i, a)\n"
+    "  real, intent(inout) :: a\n"
+    "  integer, intent(in) :: i\n"
+    "  if (i < 5) then\n"
+    "    return\n"
+    "  endif\n"
+    "  if (i > 10) then\n"
+    "    ! Comments do not matter\n"
+    "    return\n"
+    "    a=2.0 ! Dead code does not matter\n"
+    "  endif\n"
+    "  a=0.0\n"
+    "  a=1.0\n"
+    "  a=2.0\n"
+    "end subroutine\n")
+SUB_OUT1 = (
+    "subroutine sub1(i, a)\n"
+    "  real, intent(inout) :: a\n"
+    "  integer, intent(in) :: i\n\n"
+    "  if (.NOT.i < 5) then\n"
+    "    if (.NOT.i > 10) then\n"
+    "      a = 0.0\n"
+    "      a = 1.0\n"
+    "      a = 2.0\n"
+    "    end if\n"
+    "  end if\n\n"
+    "end subroutine sub1\n")
+
+# Tests with preceding code that is not part of the mask
+SUB_IN2 = (
+    "subroutine sub1(i, a)\n"
+    "  real, intent(inout) :: a\n"
+    "  integer, intent(in) :: i\n"
+    "  {0}\n"
+    "  if (i < 5) then\n"
+    "    return\n"
+    "  endif\n"
+    "  if (i > 10) then\n"
+    "    return\n"
+    "  endif\n"
+    "  a=0.0\n"
+    "end subroutine\n")
+SUB_IN2_1 = SUB_IN2.format("a = 0.0")
+SUB_IN2_2 = SUB_IN2.format(
+    "if (i > 20) then\n"
+    "    a=0.0\n"
+    "    return\n"
+    "  end if")
+SUB_IN2_3 = SUB_IN2.format(
+    "if (i > 20) then\n"
+    "    return\n"
+    "  else\n"
+    "    a=0.0\n"
+    "  end if")
+SUB_OUT2 = (
+    "subroutine sub1(i, a)\n"
+    "  real, intent(inout) :: a\n"
+    "  integer, intent(in) :: i\n\n"
+    "  {0}\n"
+    "  if (.NOT.i < 5) then\n"
+    "    if (.NOT.i > 10) then\n"
+    "      a = 0.0\n"
+    "    end if\n"
+    "  end if\n\n"
+    "end subroutine sub1\n")
+SUB_OUT2_1 = SUB_OUT2.format("a = 0.0")
+SUB_OUT2_2 = SUB_OUT2.format(
+    "if (i > 20) then\n"
+    "    a = 0.0\n"
+    "    return\n"
+    "  end if")
+SUB_OUT2_3 = SUB_OUT2.format(
+    "if (i > 20) then\n"
+    "    return\n"
+    "  else\n"
+    "    a = 0.0\n"
+    "  end if")
+
+test_cases = [(SUB_IN1, SUB_OUT1), (SUB_IN2_1, SUB_OUT2_1),
+              (SUB_IN2_2, SUB_OUT2_2), (SUB_IN2_3, SUB_OUT2_3)]
+
+
+@pytest.mark.parametrize("test_case", [0, 1, 2, 3])
```

Comment:
You could do:

    @pytest.mark.parametrize("input_code, expected", [(SUB_IN1, SUB_OUT1), (....), ...]
    def test_transformation(parser, input_code, expected):

Reply:
I had this before but then I realised that for failed tests or running "pytest -v" it was including the whole source code (without line breaks) as the name of the test. Now it looks like this:

```
=========================================================================================== test session starts ===========================================================================================
platform linux -- Python 3.8.5, pytest-6.0.1, py-1.9.0, pluggy-0.13.1 -- /home/sergi/.virtualenvs/psy3/bin/python
cachedir: .pytest_cache
rootdir: /home/sergi/workspace/stfc/PSyclone, configfile: setup.cfg
plugins: forked-1.3.0, flakes-4.0.1, pylint-0.17.0, xdist-2.1.0, cov-2.10.1, pep8-1.0.6, pep257-0.0.5
collected 6 items

src/psyclone/tests/psyir/transformations/fold_conditional_return_expressions_trans_test.py::test_description PASSED                                                                                 [ 16%]
src/psyclone/tests/psyir/transformations/fold_conditional_return_expressions_trans_test.py::test_validation PASSED                                                                                  [ 33%]
src/psyclone/tests/psyir/transformations/fold_conditional_return_expressions_trans_test.py::test_transformation[0] PASSED                                                                           [ 50%]
src/psyclone/tests/psyir/transformations/fold_conditional_return_expressions_trans_test.py::test_transformation[1] PASSED                                                                           [ 66%]
src/psyclone/tests/psyir/transformations/fold_conditional_return_expressions_trans_test.py::test_transformation[2] PASSED                                                                           [ 83%]
src/psyclone/tests/psyir/transformations/fold_conditional_return_expressions_trans_test.py::test_transformation[3] PASSED                                                                           [100%]
```
I though this was cleaner, but I can reverse it if you prefer.

Reply:
Oh I see. I don't mind much either way but I agree this output is cleaner so am happy to stick with it.